### PR TITLE
feat: add declarations attribute to js_library

### PR DESCRIPTION
**Pull request by gregmagolan**
Pre-req for https://github.com/aspect-build/rules_ts/pull/261 for the js_library with just a `.js` file case

### CODE REVIEW

File: js/private/js_library.bzl [hunk 1]
```diff
@@ -49,18 +49,27 @@ _ATTRS = {
         This includes all your checked-in code and any generated source files.
 
         The transitive npm dependencies, transitive sources & runfiles of targets in the `srcs` attribute are added to the
-        runfiles of this taregt. They should appear in the '*.runfiles' area of any executable which is output by or has a
+        runfiles of this target. They should appear in the '*.runfiles' area of any executable which is output by or has a
         runtime dependency on this target.
+
+        Source files that are JSON files, declaration files or directory artifacts will be automatically provided as
+        "declarations" available to downstream rules for type checking. To explicitly provide source files as "declarations"
+        available to downstream rules for type checking that do not match these criteria, use the `declarations`
```
Comment (alexeagle) [inline js/private/js_library.bzl:57]: "use the" -> "move those files to the"

File: js/private/js_library.bzl [hunk 2]
```diff
@@ -49,18 +49,27 @@ _ATTRS = {
...
+        attribute instead.
         """,
         allow_files = True,
     ),
+    "declarations": attr.label_list(
+        doc = """Same as `srcs` except all files are provided as "declarations" available to downstream rules for type checking.""",
```
Comment (alexeagle) [inline js/private/js_library.bzl:63]: are also provided

File: js/private/js_library.bzl [hunk 3]
```diff
@@ -163,25 +172,35 @@ target in {file_basename}'s package and add that target to the deps of {this_tar
...
 
+    sources_from_declarations, declarations = _gather_sources_and_declarations(
+        ctx = ctx,
+        targets = ctx.attr.declarations,
+        files = ctx.files.declarations,
+        is_windows = is_windows,
+    )
+
+    sources = depset(transitive = [sources, sources_from_declarations])
+    declarations = depset(transitive = [declarations, declarations_from_sources, sources_from_declarations])
```
Comment (alexeagle) [inline js/private/js_library.bzl:190]: `declarations_from_sources` alongside `sources_from_declarations` is hurting by brain (today at least) - can you bikeshed on them a bit more?

File: js/private/js_library.bzl (hunk 2 above)
Comment (alexeagle) [reply js/private/js_library.bzl:63]: and maybe give an example of why you might need that.

**Review by alexeagle — APPROVED** [review]

File: js/private/js_library.bzl (hunk 1 above)
Comment (gregmagolan) [reply js/private/js_library.bzl:57]: 👍

File: js/private/js_library.bzl (hunk 2 above)
Comment (gregmagolan) [reply js/private/js_library.bzl]: ✅

File: js/private/js_library.bzl (hunk 3 above)
Comment (gregmagolan) [reply js/private/js_library.bzl:201]: Hard to find namings that read easy here, changed to:
```
    sources, declarations = _gather_sources_and_declarations(
        ctx = ctx,
        targets = ctx.attr.srcs,
        files = ctx.files.srcs,
        is_windows = is_windows,
    )

    additional_sources, additional_declarations = _gather_sources_and_declarations(
        ctx = ctx,
        targets = ctx.attr.declarations,
        files = ctx.files.declarations,
        is_windows = is_windows,
    )

    sources = depset(transitive = [sources, additional_sources])
    declarations = depset(transitive = [declarations, additional_sources, additional_declarations])
```

**Pull request merged**